### PR TITLE
[libcurl] update usage of tools.apple_deployment_target_flag

### PR DESCRIPTION
**Pull request by SSE4**
https://github.com/conan-io/conan/pull/8263 adds `os.sdk` sub-setting, https://github.com/conan-io/conan/pull/8264 adds `os.subsystem` sub-setting, they need to be passed to the [tools.apple_deployment_target_flag](https://docs.conan.io/en/latest/reference/tools.html#tools-apple-deployment-target-flag)

P.S. the code to handle iOS in this recipe is **HUGE** mess... e.g. it sets a lot of things manually, which should have been done automatically by the `AutoToolsBuildEnvironment` helper. this has to be cleaned up, but that's for another PR (needs to be checked carefully).

Specify library name and version:  **libcurl/all**

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by a4z** [comment]
doesn't this change also mean you need a minimum conan version for being able to build that package?

File: recipes/libcurl/all/conanfile.py [hunk 1]
```diff
@@ -170,7 +170,7 @@ def build_requirements(self):
            tools.os_info.detect_windows_subsystem() != "msys2":
             self.build_requires("msys2/20200517")
         elif self._is_win_x_android:
-            self.build_requires("ninja/1.10.1")
+            self.build_requires("ninja/1.10.2")
```
Comment (SSE4) [inline recipes/libcurl/all/conanfile.py:173]: auto-added by conventions

**Comment by a4z** [comment]
great work @SSE4 , thanks for the cleanup!
Just built with the changes for iOS armv8/x86_64, and it works (not tested the result in practice , but I think that should also pass)

there are 2 more things that would be great to get rid of
```
*** Do not use buildconf. Instead, just use: autoreconf -fi
```

 and there seems to be some quirks with some path, but I am not sure if that is a curl problem, or if zlib does something wrong?

```
ld: warning: directory not found for option '-L/Users/a4z/elux/conan/zenv2/.conan/data/zlib/1.2.11/_/_/package/bc326c7d7d43e7c5f6592995fcc0db6ad7ea3b41/lib/lib'
make[1]: Nothing to be done for `all-am'.
```

**Comment by SSE4** [comment]
that part definetely looks suspicious:
```
bc326c7d7d43e7c5f6592995fcc0db6ad7ea3b41/lib/lib
```
(notice double lib)
looking at the recipe, there are some inconsistencies definitely:
https://github.com/conan-io/conan-center-index/blob/cbad6fe968139e64f94f197d6a83a6ddb157e726/recipes/libcurl/all/conanfile.py#L262
https://github.com/conan-io/conan-center-index/blob/cbad6fe968139e64f94f197d6a83a6ddb157e726/recipes/libcurl/all/conanfile.py#L257
https://github.com/conan-io/conan-center-index/blob/cbad6fe968139e64f94f197d6a83a6ddb157e726/recipes/libcurl/all/conanfile.py#L252
https://github.com/conan-io/conan-center-index/blob/cbad6fe968139e64f94f197d6a83a6ddb157e726/recipes/libcurl/all/conanfile.py#L247
https://github.com/conan-io/conan-center-index/blob/cbad6fe968139e64f94f197d6a83a6ddb157e726/recipes/libcurl/all/conanfile.py#L243

in other words, openssl, wolfssl and libnghttp2 pass `rootpath`, while zlib and libssh2 pass `lib_paths[0]`
I'll check configure options, do they require root path or lib directories

**Review by a4z — APPROVED** [review]
Great work!

**Review by prince-chrismc — APPROVED** [review]

File: recipes/libcurl/all/conanfile.py [hunk 2]
```diff
@@ -259,7 +259,7 @@ def _get_configure_command_args(self):
             params.append("--without-nghttp2")
 
         if self.options.with_zlib:
-            params.append("--with-zlib={}".format(tools.unix_path(self.deps_cpp_info["zlib"].lib_paths[0])))
+            params.append("--with-zlib={}".format(tools.unix_path(self.deps_cpp_info["zlib"].rootpath)))
```
Comment (SpaceIm) [inline recipes/libcurl/all/conanfile.py:262]: Prefer to use pkgconf: https://github.com/curl/curl/blob/f3294d9d86e6a7915a967efff2842089b8b0d071/configure.ac#L904

File: recipes/libcurl/all/conanfile.py [hunk 3]
```diff
@@ -249,7 +249,7 @@ def _get_configure_command_args(self):
             params.append("--without-wolfssl")
 
         if self.options.with_libssh2:
-            params.append("--with-libssh2={}".format(tools.unix_path(self.deps_cpp_info["libssh2"].lib_paths[0])))
+            params.append("--with-libssh2={}".format(tools.unix_path(self.deps_cpp_info["libssh2"].rootpath)))
```
Comment (SpaceIm) [inline recipes/libcurl/all/conanfile.py:252]: Same than zlib: https://github.com/curl/curl/blob/f3294d9d86e6a7915a967efff2842089b8b0d071/configure.ac#L2800

File: recipes/libcurl/all/conanfile.py [hunk 4]
```diff
@@ -347,7 +347,7 @@ def _patch_mingw_files(self):
     def _build_with_autotools(self):
         with tools.chdir(self._source_subfolder):
             # autoreconf
-            self.run("./buildconf", win_bash=tools.os_info.is_windows, run_environment=True)
+            self.run("autoreconf -fiv", win_bash=tools.os_info.is_windows, run_environment=True)
```
Comment (SpaceIm) [inline recipes/libcurl/all/conanfile.py:350]:
```suggestion
            self.run("{} -fiv".format(tools.get_env("AUTORECONF")), win_bash=tools.os_info.is_windows, run_environment=True)
```

File: recipes/libcurl/all/conanfile.py (hunk 2 above)
Comment (SSE4) [reply recipes/libcurl/all/conanfile.py:262]: I guess it's too much, it has to be tested first, and it's not the goal of the PR, so maybe for another one

File: recipes/libcurl/all/conanfile.py (hunk 4 above)
Comment (SSE4) [reply recipes/libcurl/all/conanfile.py:350]: this will raise if AUTORECONF not in the env

Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py:350]: Might need to require libtool for mingw also in build_requirement.

Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py:350]: if AUTORECONF is in env since we have autoconf in build requirement (through libtool)

**Review by a4z — DISMISSED** [review]

File: recipes/libcurl/all/conanfile.py (hunk 3 above)
Comment (a4z) [reply recipes/libcurl/all/conanfile.py:252]: not everything has to be done in this PR, SSE4 did an awesome job with this improvement,  adding more load on him is maybe not fair and nice. And there can also be space for future improvements :-)

File: recipes/libcurl/all/conanfile.py (hunk 4 above)
Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py]: build_requirement also needs to be updated, to not break MinGW:

```python
    def build_requirements(self):
        if not self._is_using_cmake_build:
            self.build_requires("libtool/2.4.6")
            self.build_requires("pkgconf/1.7.3")
            if tools.os_info.is_windows and not tools.get_env("CONAN_BASH_PATH"):
                self.build_requires("msys2/20200517")
        elif self._is_win_x_android:
            self.build_requires("ninja/1.10.2")
```

File: recipes/libcurl/all/conanfile.py [hunk 5]
```diff
@@ -170,7 +170,7 @@ def build_requirements(self):
            tools.os_info.detect_windows_subsystem() != "msys2":
             self.build_requires("msys2/20200517")
```
Comment (dmn-star) [inline recipes/libcurl/all/conanfile.py:171]: i would say that is too lax mingw detection. This leads to msys2 being installed as build requirements  even though the library is compiled for android for example. #4277

Comment (SSE4) [reply recipes/libcurl/all/conanfile.py:171]: probably, needs also `settings.os == "Windows"`? and probably, `is_mingw` property needs to check `settings.compiler == "gcc"` as well?

Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py:171]: also don't forget this comment: https://github.com/conan-io/conan-center-index/pull/4457#discussion_r572891032

File: recipes/libcurl/all/conanfile.py (hunk 4 above)
Comment (SSE4) [reply recipes/libcurl/all/conanfile.py]: I've just checked MinGW, and it fails for me:
```
libcurl/7.74.0: run_in_windows_bash: C:\.conan\d42a89\1\usr\bin\bash.exe --login -c ^"cd \^"/c/users/sse4/.conan/data/libcurl/7.74.0/_/_/build/33db161173eeaea75b0b6758cf47c091b549bebd/source_subfolder\^" ^&^& PATH=\^"/c/users/sse4/.conan/data/cmake/3.16.4/_/_/package/456f15897172eef340fcbac8a70811f2beb26a93/bin:/c/users/sse4/.conan/data/mingw_installer/1.0/conan/stable/package/d6e9d8771a7904278203ca1b715b751651fb5664/bin:/c/.conan/d42a89/1/usr/bin:/c/.conan/c5d4a7/1/bin/usr/bin:$PATH\^" ^&^& None -fiv ^"                                                                                   /usr/bin/bash: None: command not found                            
```
seems like solution to use `AUTORECONF` is not right

Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py]: You mean it fails with the modification suggested above in build requirement, or without?

Comment (SSE4) [reply recipes/libcurl/all/conanfile.py]: without
I am trying with now, but it will need to build the world (m4 and so on)

Comment (SSE4) [reply recipes/libcurl/all/conanfile.py]: fails with as well:
```
libcurl/7.74.0: run_in_windows_bash: C:\.conan\d42a89\1\usr\bin\bash.exe --login -c ^"cd \^"/c/users/sse4/.conan/data/libcurl/7.74.0/_/_/build/33db161173eeaea75b0b6758cf47c091b549bebd/source_subfolder\^" ^&^& PATH=\^"/c/users/sse4/.conan/data/cmake/3.16.4/_/_/package/456f15897172eef340fcbac8a70811f2beb26a93/bin:/c/users/sse4/.conan/data/mingw_installer/1.0/conan/stable/package/d6e9d8771a7904278203ca1b715b751651fb5664/bin:/c/.conan/d42a89/1/usr/bin:/c/users/sse4/.conan/data/libtool/2.4.6/_/_/package/e99faf16acec7281f7fd81c5696eb4133d3bc92e/bin:/c/users/sse4/.conan/data/pkgconf/1.7.3/_/_/package/6dc82da13f94df549e60f9c1ce4c5d11285a4dff/bin:/c/.conan/c5d4a7/1/bin/usr/bin:/c/users/sse4/.conan/data/automake/1.16.3/_/_/package/3e48e69237f7f2196164383ef9dedf0f93cbf249/bin:/c/users/sse4/.conan/data/autoconf/2.69/_/_/package/3475bd55b91ae904ac96fde0f106a136ab951a5e/bin:/c/users/sse4/.conan/data/m4/1.4.18/_/_/package/413a343c621037ce4d92e0bfc6f082ecc5ba0001/bin:$PATH\^" ^&^& /c/users/sse4/.conan/data/autoconf/2.69/_/_/package/3475bd55b91ae904ac96fde0f106a136ab951a5e/bin/autoreconf.exe -fiv ^"                                                                                     autoreconf.exe: Entering directory `.'                                                                                                              autoreconf.exe: configure.ac: not using Gettext                                                                                                     autoreconf.exe: running: /c/users/sse4/.conan/data/automake/1.16.3/_/_/package/3e48e69237f7f2196164383ef9dedf0f93cbf249/bin/aclocal.exe --force -I m4                                                                                                                                                   aclocal.exe: error: couldn't open directory '/c/users/sse4/.conan/data/libtool/2.4.6/_/_/package/e99faf16acec7281f7fd81c5696eb4133d3bc92e/bin/share/aclocal;/c/users/sse4/.conan/data/pkgconf/1.7.3/_/_/package/6dc82da13f94df549e60f9c1ce4c5d11285a4dff/bin/aclocal': No such file or directory        autoreconf.exe: /c/users/sse4/.conan/data/automake/1.16.3/_/_/package/3e48e69237f7f2196164383ef9dedf0f93cbf249/bin/aclocal.exe failed with exit status: 1                                                                                                                                               
```

Comment (SSE4) [reply recipes/libcurl/all/conanfile.py]: okay, it works **WITHOUT** `libtool/2.4.6` and `pkgconf/1.7.3` in build requirements, with the following fix:
```
self.run("{} -fiv".format(tools.get_env("AUTORECONF") or "autoreconf"), win_bash=tools.os_info.is_windows, run_environment=True)
```
it turns out, MinGW has `autoreconf` which works out of the box, so it doesn't need build requirement (?)
should I commit these changes then? 
I am not sure, MinGW support goes outside of the scope of the PR. 
so, probably just leave it as it was before (calling `./buildconf`)?

Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py]: it's fine for me (keep buildconf, and replace this logic in another PR).

Just one comment, and @madebr may confirm: msys2 recipe does not guarantee to provide autoreconf of pkg-config.

Comment (SSE4) [reply recipes/libcurl/all/conanfile.py]: @a4z I probably will revert back to `buildconf` for now, as it seems to be problematic to avoid breakage for MinGW. we may try one more time in another PR :)

File: recipes/libcurl/all/conanfile.py [hunk 6]
```diff
@@ -347,7 +347,7 @@ def _patch_mingw_files(self):
     def _build_with_autotools(self):
         with tools.chdir(self._source_subfolder):
             # autoreconf
-            self.run("./buildconf", win_bash=tools.os_info.is_windows, run_environment=True)
+            self.run("{} -fiv".format(tools.get_env("AUTORECONF")), win_bash=tools.os_info.is_windows, run_environment=True)
```
Comment (SSE4) [inline recipes/libcurl/all/conanfile.py:350]:
```suggestion
            self.run("./buildconf", win_bash=tools.os_info.is_windows, run_environment=True)
```
to hit the button tomorrow morning, if we decide

File: recipes/libcurl/all/conanfile.py (hunk 4 above)
Comment (madebr) [reply recipes/libcurl/all/conanfile.py]: See https://github.com/curl/curl/blob/ec5d9b44a2e837fc7b82d1c60d5fae3f851620dc/buildconf#L3
I don't see the point in using buildconf

Comment (SSE4) [reply recipes/libcurl/all/conanfile.py]: as I understand, `buildconf` script is more or less equivalent to:
```
self.run("{} -fiv".format(tools.get_env("AUTORECONF") or "autoreconf"), win_bash=tools.os_info.is_windows, run_environment=True)
```
so, this change should be okay

File: recipes/libcurl/all/conanfile.py (hunk 6 above)
Comment (SSE4) [inline recipes/libcurl/all/conanfile.py:350]:
```suggestion
            self.run("{} -fiv".format(tools.get_env("AUTORECONF") or "autoreconf"), win_bash=tools.os_info.is_windows, run_environment=True)
```

File: recipes/libcurl/all/conanfile.py (hunk 5 above)
Comment (dmn-star) [reply recipes/libcurl/all/conanfile.py:171]: @SSE4 I guess so.
https://github.com/conan-io/conan-center-index/blob/4cfe106fab3bf1593d39f802a7a5a69ee68250b3/recipes/libcurl/all/conanfile.py#L77

 return self.settings.os == "Windows" and self.settings.compiler == "gcc"

Comment (dmn-star) [reply recipes/libcurl/all/conanfile.py:171]: @SpaceIm _is_using_cmake_build is always true when cross compile for Android on Windows

        if not self._is_using_cmake_build:
        ....
        elif self._is_win_x_android:

https://github.com/conan-io/conan-center-index/blob/4cfe106fab3bf1593d39f802a7a5a69ee68250b3/recipes/libcurl/all/conanfile.py#L84-L85

Comment (SSE4) [reply recipes/libcurl/all/conanfile.py:171]: there is a couple of another weirdness in the recipe:
- CMake build is used only for some platforms (e.g. Android)
- Android build enforces ninja (shouldn't it work with any generator?)

anyway, I am not going to address this right now. let's just merge what we have already, otherwise, it will never end :) future PR may address these problems.

**Comment by SSE4** [comment]
okay, let's finalize, no more improvements, please :) as every new change has to be tested with MinGW and iOS (which are not covered by CCI), which is not very convenient for me (it requires a reboot to bootcamp and back to mac).

what is not included (and won't be) in this PR:
- switching zlib and other deps to pkg-config
- using conan's autoconf for MinGW (appears to be broken)
- not using MSYS for Android

**Review by a4z — APPROVED** [review]

**Comment by a4z** [comment]
Awesome work!  
and I totally agree that the rest has to be part of future iterations.

File: recipes/libcurl/all/conanfile.py (hunk 5 above)
Comment (dmn-star) [reply recipes/libcurl/all/conanfile.py:171]: No problem, building for Android on Windows is just my hobby project. :-) However what I have never understood why the autoconfig build is still used when the CMake build is available.

**Review by dmn-star — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]
Merge!

File: recipes/libcurl/all/conanfile.py (hunk 4 above)
Comment (dmn-star) [reply recipes/libcurl/all/conanfile.py]: What do you need autoreconf for here anyway? Wouldn't configure be enough?

**Review by SpaceIm — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
